### PR TITLE
SetEnvelopeName() and GetEnvelopeName() extension methods

### PR DESCRIPTION
**Pull request by Dmitry-Matveev**
@SergeyKanzhelev , added `SetEnvelopeName()` and  `GetEnvelopeName()` extension methods as we discussed offline.

New extension method allows to override default envelope name for the serialized telemetry.
For instance, one can replace `Request` envelope name on the wire with `Request.Light`, so that processing can be aware that this is not basic request.

~~Note 1: This is first draft, many design questions are still unanswered.
Note 2: I'm having issues with tests discovery in local VS, will fix tests failing in the pipeline.~~

### Notes for reviewers:

- We support [comment build triggers](https://docs.microsoft.com/azure/devops/pipelines/repos/github?view=azure-devops&tabs=yaml#comment-triggers)
  - `/AzurePipelines run` will queue all builds
  - `/AzurePipelines run <pipeline-name>` will queue a specific build

### CODE REVIEW

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/EventTelemetry.cs [hunk 1]
```diff
@@ -15,9 +15,9 @@
     /// </summary>
     public sealed class EventTelemetry : ITelemetry, ISupportProperties, ISupportAdvancedSampling, ISupportMetrics, IAiSerializableTelemetry
     {
-        internal const string TelemetryName = "Event";
-
+        internal const string DefaultEnvelopeName = "Event";
```
Comment (Dmitry-Matveev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/EventTelemetry.cs]: This was the only event type that had previous internal const exposed to the pipeline. Preserving it as internal const for now - let me know if you'd like to refactor this.

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Telemetry.cs [hunk 2]
```diff
@@ -32,12 +32,11 @@ public static string WriteTelemetryName(this ITelemetry telemetry, string teleme
         {
             // A different event name prefix is sent for normal mode and developer mode.
             // Format the event name using the following format:
-            // Microsoft.ApplicationInsights[.Dev].<normalized-instrumentation-key>.<event-type>
```
Comment (Dmitry-Matveev) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Telemetry.cs]: Are we certain that no existing experiences depend on the envelope name being formatted as "Microsoft.ApplicationInsights[.Dev].IKey.EventName", e.g. VS integration or anything in the pipeline?

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/TelemetryExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        /// <exception cref="ArgumentException">Concrete implementation of ITelemetry object does not expose envelope name.</exception>
+        public static void SetEnvelopeName(this ITelemetry telemetry, string envelopeName)
+        {
+            if (telemetry is IAiSerializableTelemetry aiSerializableTelemetry)
+            {
+                aiSerializableTelemetry.TelemetryName = envelopeName;
+            }
+            else
+            {
+                throw new ArgumentException("Provided telemetry object does not support envelope name.", nameof(telemetry));
```
Comment (Dmitry-Matveev) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/TelemetryExtensions.cs]: Not a big fan of this throw, but currently only `IAiSerializableTelemetry` exposes envelope name and it's an internal-only interface that was meant to remain internal. Currently, any out-of-the-box `ITelemetry` implements `IAiSerializableTelemetry`, but that may not be the case if someone implements custom `ITelemetry` or we later ship non-serializable type. Thoughts?

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Telemetry.cs (hunk 2 above)
Comment (SergeyKanzhelev) [reply BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Telemetry.cs]: I think whole business of envelope names was for the internal customers ingesting data via Vortex. I'd suggest to eliminate this all together and simply use EnvelopeName. If anybody wants to do string formatting - they can write a telemetry initializer

File: .publicApi/Microsoft.ApplicationInsights.dll/netstandard2.0/PublicAPI.Unshipped.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Microsoft.ApplicationInsights.Extensibility.Implementation.TelemetryExtensions
+static Microsoft.ApplicationInsights.Extensibility.Implementation.TelemetryExtensions.SetEnvelopeName(this Microsoft.ApplicationInsights.Channel.ITelemetry telemetry, string envelopeName) -> void
```
Comment (SergeyKanzhelev) [inline .publicApi/Microsoft.ApplicationInsights.dll/netstandard2.0/PublicAPI.Unshipped.txt]: we also need Get so GenevaChannel will be able to use it as an event name

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/TelemetryExtensions.cs (hunk 3 above)
Comment (SergeyKanzhelev) [reply BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/TelemetryExtensions.cs]: Just use default `"Event"`

File: BASE/src/Microsoft.ApplicationInsights/Constants.cs [hunk 5]
```diff
@@ -4,9 +4,9 @@ internal class Constants
     {
         internal const string ProfileQueryEndpoint = "https://dc.services.visualstudio.com/api/profiles/{0}/appId";
 
-        internal const string TelemetryNamePrefix = "Microsoft.ApplicationInsights.";
```
Comment (TimothyMothra) [inline BASE/src/Microsoft.ApplicationInsights/Constants.cs]: do we need to keep this empty variable?

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/RichPayloadEventSource.cs [hunk 6]
```diff
@@ -259,7 +259,7 @@ public void Process(ITelemetry item)
                 CopyGlobalPropertiesIfRequired(item, telemetryItem.Properties);
                 item.Sanitize();
                 this.WriteEvent(
-                    AvailabilityTelemetry.TelemetryName,
+                    telemetryItem.EnvelopeName,
                     telemetryItem.Context.InstrumentationKey,
                     telemetryItem.Context.SanitizedTags,
                     telemetryItem.Data,
```
Comment (TimothyMothra) [inline BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/RichPayloadEventSource.cs]: would be great if this giant if/else could be consolidated into a single interface. :)
The only thing unique that i see is the Keyword, which could be put on the TelemetryItem as an internal field.

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 7]
```diff
@@ -2,24 +2,24 @@
 {
     public class ItemType
     {
-        public const string Metric = "Microsoft.ApplicationInsights.Metric";
+        public const string Metric = "Metric";
 
-        public const string Request = "Microsoft.ApplicationInsights.Request";
+        public const string Request = "Request";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string Request = "AppRequests";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 8]
```diff
@@ -2,24 +2,24 @@
 {
     public class ItemType
     {
-        public const string Metric = "Microsoft.ApplicationInsights.Metric";
+        public const string Metric = "Metric";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string Metric = "AppMetrics";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 9]
```diff
@@ -2,24 +2,24 @@
...
 
-        public const string Request = "Microsoft.ApplicationInsights.Request";
+        public const string Request = "Request";
 
-        public const string Exception = "Microsoft.ApplicationInsights.Exception";
+        public const string Exception = "Exception";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string Exception = "AppExceptions";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 10]
```diff
@@ -2,24 +2,24 @@
...
 
-        public const string Message = "Microsoft.ApplicationInsights.Message";
+        public const string Message = "Message";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string Message = "AppTraces";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 11]
```diff
@@ -2,24 +2,24 @@
...
 
-        public const string Event = "Microsoft.ApplicationInsights.Event";
+        public const string Event = "Event";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string Event = "AppEvents";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 12]
```diff
@@ -2,24 +2,24 @@
...
 
-        public const string PageView = "Microsoft.ApplicationInsights.PageView";
+        public const string PageView = "PageView";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string PageView = "AppPageViews";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 13]
```diff
@@ -2,24 +2,24 @@
...
 
-        public const string PageViewPerformance = "Microsoft.ApplicationInsights.PageViewPerformance";
+        public const string PageViewPerformance = "PageViewPerformance";
 
-        public const string RemoteDependency = "Microsoft.ApplicationInsights.RemoteDependency";
+        public const string RemoteDependency = "RemoteDependency";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string RemoteDependency = "AppDependencies";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 14]
```diff
@@ -2,24 +2,24 @@
...
 
-        public const string PerformanceCounter = "Microsoft.ApplicationInsights.PerformanceCounter";
+        public const string PerformanceCounter = "PerformanceCounter";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string PerformanceCounter = "AppPerformanceCounters";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 15]
```diff
@@ -2,24 +2,24 @@
...
 
-        public const string SessionState = "Microsoft.ApplicationInsights.SessionState";
+        public const string SessionState = "SessionState";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]: is it even used anywhere? Shouldn't. Let's remove

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 16]
```diff
@@ -2,24 +2,24 @@
...
+        public const string Message = "Message";
 
-        public const string Event = "Microsoft.ApplicationInsights.Event";
+        public const string Event = "Event";
 
-        public const string PageView = "Microsoft.ApplicationInsights.PageView";
+        public const string PageView = "PageView";
 
-        public const string PageViewPerformance = "Microsoft.ApplicationInsights.PageViewPerformance";
+        public const string PageViewPerformance = "PageViewPerformance";
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]:
```suggestion
        public const string PageViewPerformance = "AppBrowserTimings";
```

File: BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs [hunk 17]
```diff
@@ -2,24 +2,24 @@
 {
     public class ItemType
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/ApplicationInsightsTypes/ItemType.cs]: using the new (upcoming) table names

File: BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Channel/ITelemetryTest.cs [hunk 18]
```diff
@@ -222,7 +222,7 @@ private void SerializeWritesTelemetryNameAsExpectedByEndpoint()
                 .SerializeDeserializeTelemetryItem<TEndpointData>(expected);
 
             Assert.AreEqual(
-                Constants.TelemetryNamePrefix + "312cbd799dbb4c48a7da3cc2a931cb71." + this.ExtractTelemetryNameFromType(typeof(TTelemetry)),
+                this.ExtractTelemetryNameFromType(typeof(TTelemetry)),
```
Comment (SergeyKanzhelev) [inline BASE/Test/Microsoft.ApplicationInsights.Test/Microsoft.ApplicationInsights.Tests/Channel/ITelemetryTest.cs]: this test will fail... perhaps it is not needed

File: BASE/src/Microsoft.ApplicationInsights/Constants.cs [hunk 19]
```diff
@@ -4,9 +4,7 @@ internal class Constants
...
-
-        internal const string DevModeTelemetryNamePrefix = "Microsoft.ApplicationInsights.Dev.";
+        internal const string DevModeTelemetryNamePrefix = "Dev.";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/Constants.cs]: I don't think this logic is needed. It was introduced for Vortex customers and they will need to build a telemetry initializer now anyway. So let's just remove this extra thing from the base SDK.

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/AvailabilityTelemetry.cs [hunk 20]
```diff
@@ -15,9 +15,8 @@
     /// </summary>
     public sealed class AvailabilityTelemetry : ITelemetry, ISupportProperties, ISupportMetrics, IAiSerializableTelemetry
     {
-        internal const string TelemetryName = "Availability";
-
         internal readonly AvailabilityData Data;
+        internal string EnvelopeName = "Availability";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/AvailabilityTelemetry.cs]:
```suggestion
        internal string EnvelopeName = "AppAvailabilityResults";
```

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/DependencyTelemetry.cs [hunk 21]
```diff
@@ -19,7 +19,7 @@ namespace Microsoft.ApplicationInsights.DataContracts
     /// </summary>
     public sealed class DependencyTelemetry : OperationTelemetry, ITelemetry, ISupportProperties, ISupportAdvancedSampling, ISupportMetrics, IAiSerializableTelemetry
     {
-        internal new const string TelemetryName = "RemoteDependency";
+        internal string EnvelopeName = "RemoteDependency";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/DependencyTelemetry.cs]:
```suggestion
        internal string EnvelopeName = "AppDependencies";
```

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/EventTelemetry.cs (hunk 1 above)
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/EventTelemetry.cs]:
```suggestion
        internal const string DefaultEnvelopeName = "AppEvents";
```

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/ExceptionTelemetry.cs [hunk 22]
```diff
@@ -18,7 +18,7 @@
     /// </remarks>
     public sealed class ExceptionTelemetry : ITelemetry, ISupportProperties, ISupportAdvancedSampling, ISupportMetrics, IAiSerializableTelemetry
     {
-        internal const string TelemetryName = "Exception";
+        internal string EnvelopeName = "Exception";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/ExceptionTelemetry.cs]:
```suggestion
        internal string EnvelopeName = "AppExceptions";
```

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/MetricTelemetry.cs [hunk 23]
```diff
@@ -15,10 +15,9 @@
     /// </summary>
     public sealed class MetricTelemetry : ITelemetry, ISupportProperties, IAiSerializableTelemetry
     {
-        internal const string TelemetryName = "Metric";
-
         internal readonly MetricData Data;
         internal readonly DataPoint Metric;
+        internal string EnvelopeName = "Metric";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/MetricTelemetry.cs]:
```suggestion
        internal string EnvelopeName = "AppMetrics";
```

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/PageViewPerformanceTelemetry.cs [hunk 24]
```diff
@@ -13,9 +13,8 @@
     /// </summary>
     public sealed class PageViewPerformanceTelemetry : ITelemetry, ISupportProperties, ISupportAdvancedSampling, IAiSerializableTelemetry
     {
-        internal const string TelemetryName = "PageViewPerformance";
-
         internal readonly PageViewPerfData Data;
+        internal string EnvelopeName = "PageViewPerformance";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/PageViewPerformanceTelemetry.cs]:
```suggestion
        internal string EnvelopeName = "AppBrowserTimings";
```

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/PageViewTelemetry.cs [hunk 25]
```diff
@@ -20,9 +20,8 @@
     /// </remarks>
     public sealed class PageViewTelemetry : ITelemetry, ISupportProperties, ISupportAdvancedSampling, ISupportMetrics, IAiSerializableTelemetry
     {
-        internal const string TelemetryName = "PageView";
-
         internal readonly PageViewData Data;
+        internal string EnvelopeName = "PageView";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/PageViewTelemetry.cs]:
```suggestion
        internal string EnvelopeName = "AppPageViews";
```

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/RequestTelemetry.cs [hunk 26]
```diff
@@ -22,7 +22,7 @@
     /// </remarks>
     public sealed class RequestTelemetry : OperationTelemetry, ITelemetry, ISupportProperties, ISupportMetrics, ISupportAdvancedSampling, IAiSerializableTelemetry
     {
-        internal new const string TelemetryName = "Request";
+        internal string EnvelopeName = "Request";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/RequestTelemetry.cs]:
```suggestion
        internal string EnvelopeName = "AppRequests";
```

File: BASE/src/Microsoft.ApplicationInsights/DataContracts/TraceTelemetry.cs [hunk 27]
```diff
@@ -15,9 +15,8 @@
     /// </summary>
     public sealed class TraceTelemetry : ITelemetry, ISupportProperties, ISupportAdvancedSampling, IAiSerializableTelemetry
     {
-        internal const string TelemetryName = "Message";
-
         internal readonly MessageData Data;
+        internal string EnvelopeName = "Message";
```
Comment (SergeyKanzhelev) [inline BASE/src/Microsoft.ApplicationInsights/DataContracts/TraceTelemetry.cs]:
```suggestion
        internal string EnvelopeName = "AppTraces";
```

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Telemetry.cs (hunk 2 above)
Comment (SergeyKanzhelev) [reply BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/Telemetry.cs]: this prefix as well as ikey was introduced for internal Vortex customers. Let's remove it as they will need to create telemetry initializer anyway. So they can take care of it in that initializer

File: BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/TelemetryExtensions.cs (hunk 3 above)
Comment (SergeyKanzhelev) [reply BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/TelemetryExtensions.cs]: I see. I'd suggest using the default name here as well.

Comment (SergeyKanzhelev) [reply BASE/src/Microsoft.ApplicationInsights/Extensibility/Implementation/TelemetryExtensions.cs]: really don't want exceptions in serialization

File: BASE/src/Microsoft.ApplicationInsights/Constants.cs (hunk 19 above)
Comment (TimothyMothra) [reply BASE/src/Microsoft.ApplicationInsights/Constants.cs]: @SergeyKanzhelev do you have a contact on that team? I think we should communicate this change.

Comment (SergeyKanzhelev) [reply BASE/src/Microsoft.ApplicationInsights/Constants.cs]: It's all Vortex users. I don't have contacts, but Tillman should know

**Comment by TimothyMothra** [comment]
@Dmitry-Matveev @SergeyKanzhelev This is available now, 2.14.0-beta3

**Pull request merged**